### ISSUE TITLE
Use airlock processes to handle the operations of an airlock

### ISSUE THREAD

**Issue by Urwumpe** [mars-sim/mars-sim#241]
Instead of defining a simple state in the Airlock function of a building, every building with one or more airlock functions should be able to execute an an airlock process similar to the ManufacturingProcess for Manufacturing. 

An airlock should have at four states: IDLE, INGRESS, EGRESS and DEFECT.

Every new airlock process gets initially placed into a "kindness" queue, waiting a predefined long enough period of time for enough settlers to join the process. When enough settlers are around to fill the airlock, the process starts immediately. 

There can only be one active process per airlock. 

The airlock function of a building should keep track of air lock waiting queues for entering and leaving settlers. 

There should be two major airlock processes: IngressByAirlock and EgressByAirlock.  

For each such process, the first settler to initiate the process will be the operator of the airlock (unless a different process gets defined and used there). 

When entering a building from outside (ingress), first the queue waiting outside is checked. Settlers who decide to enter the settlement will prefer a building with an already scheduled ingress process. First of all, the inner door is closed and pressure is equalized to outside. When this is achieved (or already done before), the airlock opens its outside door and the settlers assigned to the current process will enter. Then the outside door gets closed and after that, the pressure gets equalized to the inside pressure. After reaching habitable conditions, the settlers doff their EVA suits inside the airlock, store them in the airlock (or the settlement inventory) and enter the building through the inner door. After that, the process ends and the airlock is available for a new process. 

Egress operates opposite.

Since there can be no depressurization of the airlock with settlers still without EVA suit, this should be much safer.

A third process "AirlockIVA" allows to equalize pressure to the internal settlement pressure in the airlock for maintenance and accessing it. The process ends with the airlock at internal pressure, the inner door open and the outer door closed. 

The state of an airlock process should be presented in the UI like a manufacturing process with a progress bar, the pressure inside the airlock, the list of settlers in the airlock  (the operator could be highlighted) and the display name of the process: "Ingress"/"Egress"/"IVA".

The queues of settlers waiting at a building could be represented by two lists of settler names.

**Comment by mokun** [comment]
> There should be two major airlock processes: IngressByAirlock and EgressByAirlock.

Yes they are sort of implicit in the airlock operation as we have been relying on concrete class `BuildingAirlock` and `VehicleAirlock` that extends the abstract class `Airlock` as the mechanism for ingress and egress of the surface EVA.

https://github.com/mars-sim/mars-sim/blob/c19700ce0aa63aaf4d3bff0ff6822c2db85e8095/mars-sim-core/src/main/java/org/mars_sim/msp/core/structure/Airlock.java#L60-L86

> An airlock should have at four states: IDLE, INGRESS, EGRESS and DEFECT.

Previous developers did model the airlock in 4 states with enum `AirlockState`, namely,
*PRESSURIZED, DEPRESSURIZED, PRESSURIZING, & DEPRESSURIZING*.

Yea we could add more states. 

Say if this `EVA Airlock` building has malfunction, I wonder if the MalfunctionManager would prevent the airlock from being operated and what should happen to the occupants right in the middle of the actions ? 

Besides, tracking who are inside the airlock, each airlock would typically elect someone whose most experience in EVA ops as the *Airlock Operator*.

Currently it also models the opening and closing of the inner door and outer door for proper ingress and egress.

Having said that I have hoped that we have pretty much accounted for most (if not all) possible logical actions and fixed all the known airlock bugs :D

**Comment by Urwumpe** [comment]
Hmm... there should be some malfunction process to recover the settlers then if the airlock fails in the middle of the process... maybe "AbortIngress" when the outer door is still openable and "AbortEgress" if the inner door can still be opened. And "AbortIngress" could also work when both doors have failed open. 

Repairing the inner door should be possible from inside the building...

**Comment by mokun** [comment]
I just ran into a situation as follows : 

```
00-Adir-14:473.671 (Severe) Inventory::canStoreUnit : Carolyn Harris had a mass of 164.0 kg and was too big for the owner 'Dune Runner' to carry. Remaining Cap : 85.5 kg. Gen Cap : 2500.0 kg.
00-Adir-14:473.671 (Severe) TaskManager : Carolyn Harris had trouble calling performTask(). [ERROR: UNUSED LOG ARGUMENTS]
java.lang.IllegalStateException: Unit: Carolyn Harris could not be stored in Dune Runner
	at org.mars_sim.msp.core.Inventory.transferUnit(Inventory.java:2405)

```

**Comment by Urwumpe** [comment]
Maybe we should include a mass budget plan into a mission plan for approval in first place to prevent such situations. At least I would find it very interesting to read, how much regolith could be carried aboard an exploration rover at all. I suspect, with the consumables for the return leg, it is just enough for rather scientific samples than a large amont of ore for industry. For example, there could be one such budget for each waypoint and a calculation of the changes:

1. A list of the changes when leaving a waypoint (loading the rover with goods in the settlement or collected ore at the exploration site)
2. Travelled distance vs planned distance to waypoint
3. Real time to waypoint vs planned time to waypoint
4. Maybe as extension for later, the name of the known route to a waypoint from a previous mission to it - if there is one
5.  A list of the planned consumption during the travel to the next waypoint (oxygen, methane, food) and the real values during the travel. 
6. At the final waypoint, there should be a mission end mass budget consisting only of point 1 for the offloading of the rover there.
7. Additionally, there could be a mission total mass budget summarizing the total distances, (all routes used or explored), all inventory changes (If as much was loaded at the destination base as was offloaded after the mission, this should be zero).

But the safest option would be tolerating overloading a rover a bit, instead of having a hard limit. I suggest we add some penalty function there to discourage exceeding the general capacity too much:

- No effects until 110% general capacity (safety margin)
- Increased chance of malfunctions from 120% on, reaching the double chance of malfunction for certain parts already at 150% general capacity
- Maximum possible speed gets reduced from 100% general capacity on, reaching zero at 200% general capacity.  (which means, loading an exploration rover to 3750 kg would half its top speed, but still permit travel)

**Comment by mokun** [comment]
>  include a mass budget plan into a mission plan for approval in first place 

Yes we can tweak the initial condition as to how the mission plan gets approved.

https://github.com/mars-sim/mars-sim/blob/90869283c2ec51689102142ab32fada49a40240f/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/mission/CollectRegolith.java#L41-L48

From `CollectRegolith` above, you can see it roughly set the limit to 2000 kg.

But there are a few unknown. 

For instance, the weight of all the occupants are not known initially until after they are being selected. 

How much consumables will be onboard the rover is not known until they are being loaded.

They are actually no big deal since we can write a few methods to keep track of them.

Will add more safeguard to prevent silly logical mistakes like that :D 

> But the safest option would be tolerating overloading a rover a bit, instead of having a hard limit. I suggest we add some penalty function there to discourage exceeding the general capacity too much:

Sure. But we'll need to figure how to circumvent this hard limit things in the `Inventory` class since the general capacity is set to a decimal number.

**Comment by mokun** [comment]
> Maximum possible speed gets reduced from 100% general capacity on, reaching zero at 200% general capacity. (which means, loading an exploration rover to 3750 kg would half its top speed, but still permit travel)

To remove the hard limit on the general capacity in `Inventory`, it seems big changes are needed. 

So I'm modifying how a person or robot's walking speed would be reduced, based upon the load one is carrying. 

Will commit the build soon.

**Comment by mokun** [comment]
Closing this now.

We've since refined the processes in the airlock. So far the logics in each task phase are still very prone to error and a person can get stuck.

Perhaps we'll have to wait for the concept of having a series of `Simple Task` (that will have defensive programming mechanism built-in) be implemented so that we can tackle on this airlock again.

If anyone of us are interested in revamping the ingress and egress of an airlock, do use the brand new GitHub project interface to create a project. See https://github.com/mars-sim/mars-sim/projects?query=is%3Aopen.

Also, we refine the carrying capacity of a person.

**Comment by mokun** [comment]
@bevans2000 

If anyone of us are interested in revamping the ingress and egress of an airlock, do use the brand new GitHub project interface to create a project. See https://github.com/mars-sim/mars-sim/projects?query=is%3Aopen.
